sdks/go/pkg/beam/core/runtime/graphx/xlang.go

```diff
@@ -16,11 +16,222 @@
+		}
+		if tag == graph.UnnamedOutputTag && len(expandedOutputs) > 1 {
+			panic(errors.Errorf("mismatched number of unnamed outputs:\nreceived - %v\nexpected - 1", len(expandedOutputs)))
+		}
+	}
+}
+
+// ResolveOutputIsBounded updates each Output node with respect to the received
+// expanded components to reflect if it is bounded or not
+func ResolveOutputIsBounded(e *graph.MultiEdge, isBoundedUpdater func(*graph.Node, bool)) {
+	ext := e.External
+	exp := ext.Expanded
+	components, err := ExpandedComponents(exp)
+	if err != nil {
+		panic(err)
+	}
+	expandedPCollections := components.GetPcollections()
+	transform, err := ExpandedTransform(exp)
+	if err != nil {
+		panic(err)
+	}
+	expandedOutputs := transform.GetOutputs()
+
+	for tag, node := range ExternalOutputs(e) {
+		var id string
+		isBounded := true
+
+		switch tag {
+		case graph.UnnamedOutputTag:
+			for _, id = range expandedOutputs {
+				// easiest way to access map with one entry (key,value)
+			}
+		default:
+			id = expandedOutputs[tag]
+		}
+
+		if pcol, exists := expandedPCollections[id]; exists {
+			if pcol.GetIsBounded() == pipepb.IsBounded_UNBOUNDED {
+				isBounded = false
+			}
+			isBoundedUpdater(node, isBounded)
+		} else {
+			panic(errors.Errorf("missing corresponsing pcollection of named output: %v is expected in %v", id, expandedPCollections))
+		}
+
+	}
+}
+
+// AddFakeImpulses adds an impulse transform as the producer for each input to
+// the root transform. Inputs need producers to form a correct pipeline.
+func AddFakeImpulses(p *pipepb.Pipeline) {
+	// For a pipeline consisting of only the external node edge, there will be
+	// single root transform which will be the external transform.
+	// Adding fake impulses per input to this external transform
+	transforms := p.GetComponents().GetTransforms()
+	ext := transforms[p.GetRootTransformIds()[0]]
+
+	for tag, id := range ext.GetInputs() {
+		key := fmt.Sprintf("%s_%s", "impulse", tag)
+		output := map[string]string{"out": id}
+
+		impulse := &pipepb.PTransform{
+			UniqueName: key,
+			Spec: &pipepb.FunctionSpec{
+				Urn: URNImpulse,
+			},
+			Outputs: output,
+		}
+
+		transforms[key] = impulse
+	}
+
+}
+
+// RemoveFakeImpulses removes each fake impulse per input to the the transform.
+// Multiple producers for one Input cannot be present.
+func RemoveFakeImpulses(c *pipepb.Components, ext *pipepb.PTransform) {
+	transforms := c.GetTransforms()
+	var impulseIDs []string
+
+	for tag := range ext.GetInputs() {
+		id := fmt.Sprintf("%s_%s", "impulse", tag)
+		impulseIDs = append(impulseIDs, id)
+	}
+
+	for _, id := range impulseIDs {
+		t := transforms[id]
+		if t.GetSpec().GetUrn() == URNImpulse {
+			delete(transforms, id)
+
+		}
+	}
+}
+
 // ExpandedComponents type asserts the Components field with interface{} type
 // and returns its pipeline component proto representation
 func ExpandedComponents(exp *graph.ExpandedTransform) (*pipepb.Components, error) {
```

Comment:
Can this function be unexported now that callers were moved into graphx?

Reply:
This one specifically is still used outside the package (in `xlangx.ResolveArtifacts`). The `ExternalOutputs` helper function is also used outside the package. And since some of them are used outside, I feel like all four of these helpers should be exported, just for consistency.